File: examples/example-ios/polar-sensor-data-collector/PSDC/Views/Graph/HrDataHolder.swift
//  Copyright © 2026 Polar. All rights reserved.

import Foundation

struct HrSample {
    let hr: Int
}

struct HrState {
    var hrSamples: [HrSample] = []
    var currentHr: Int = 0
}

class HrDataHolder: ObservableObject {
    static let shared = HrDataHolder()
    
    private static let maxHrSamples = 300
    
    @Published private(set) var hrState = HrState()
    
    private var hrSamplesList: [HrSample] = []
    
    private init() {}
    
    func updateHr(_ hr: Int) {
        let sample = HrSample(hr: hr)
        hrSamplesList.append(sample)
        
        hrState = HrState(
            hrSamples: hrSamplesList,
            currentHr: hr
        )
        
        if hrSamplesList.count > Self.maxHrSamples {
            hrSamplesList.removeFirst()
        }
    }
    
    func clear() {
        hrSamplesList.removeAll()
        hrState = HrState()
    }
}

File: examples/example-android/polar-sensor-data-collector/app/src/main/java/com/polar/polarsensordatacollector/ui/graph/HrDataHolder.kt
package com.polar.polarsensordatacollector.ui.graph

import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object HrDataHolder {
    private const val TAG = "HrDataHolder"

    data class HrSample(
        val hr: Int,
    )

    data class HrState(
        val hrSamples: List<HrSample> = emptyList(),
        val currentHr: Int = 0,
    )

    private val _hrState = MutableStateFlow(HrState())
    val hrState: StateFlow<HrState> = _hrState.asStateFlow()

    private val hrSamplesList = mutableListOf<HrSample>()

    fun updateHr(hr: Int) {
        val sample = HrSample(hr)
        hrSamplesList.add(sample)

        val newState = HrState(
            hrSamples = hrSamplesList.toList(),
            currentHr = hr
        )
        _hrState.value = newState
    }

    fun clear() {
        hrSamplesList.clear()
        _hrState.value = HrState()
    }
}